File: cpp_solutions/Array/Search_for_a_Range.cpp
/*
*
* Search for a range
* Given an array of integers sorted in ascending order, find the starting and ending position of a given target value.
* 
* Notes: binary search
* 
**/

#include <vector>
using namespace std;

vector<int> searchRange(vector<int>& nums, int target){
	vector<int> rst;
	// rule out edge cases
	if(nums.size() == 0){
		rst.push_back(-1);
		rst.push_back(-1);
		return rst;
	}

	int leftBound = searchLeftBound(nums, target);
	int rightBound =  searchRightBound(nums, target);

	rst.push_back(leftBound);
	rst.push_back(rightBound);
	return rst;
}

int searchLeftBound(vector<int>& nums, int target){
	int start = 0, end = nums.size() - 1;
	while(start + 1 < end){
		int mid = (start + end) / 2;

		if(nums[mid] == target){
			end = mid;
		}
		else if(nums[mid] < target){
			start = mid;
		}
		else{
			end = mid;
		}
	}

	if(nums[start] == target){ return start; }
	else if(nums[end] == target) {return end;}
	else {return -1; }
}

int searchRightBound(vector<int>& nums, int target){
	int start = 0, end = nums.size() - 1;
	while(start + 1 < end){
		int mid = (start + end) / 2;

		if(nums[mid] == target){
			start = mid;
		}
		else if(nums[mid] < target){
			start = mid;
		}
		else{
			end = mid;
		}
	}
	
	if(nums[end] == target) {return end;}
	else if(nums[start] == target){ return start; }
	else {return -1; }
}

File: python_solutions/Array/Search_for_a_Range.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ret = [-1, -1]
        if len(nums) == 0:
            return ret
        
        start, end = 0, len(nums)-1
        while start+1 < end:
            mid = start+int((end-start)/2)
            if nums[mid] >= target:
                end = mid
            else:
                start = mid
                
        if nums[start] == target:
            ret[0] = start
        elif nums[end] == target:
            ret[0] = end
        else:
            return ret
        
        start, end = ret[0], len(nums)-1
        while start+1 < end:
            mid = start+int((end-start)/2)
            if nums[mid] <= target:
                start = mid
            else:
                end = mid
        
        if nums[end] == target:
            ret[1] = end
        elif nums[start] == target:
            ret[1] = start
        else:
            ret[1] = ret[0]
            
        return ret
                
# Notes: run binary search twice
